### PR TITLE
Add NMS warmup for clearer post-processing latency

### PR DESCRIPTION
**Pull request by Y-T-G**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds configurable pin_memory to dataloaders and warms up NMS during backend warmup to smooth first-iteration latency and give users better memory control. ⚡🧠

### 📊 Key Changes
- build_dataloader now accepts a new parameter: `pin_memory: bool = True`.
  - Internally uses `pin_memory = (num_device_gpus > 0) and pin_memory`.
- Validation dataloader sets `pin_memory=self.training`, disabling pinning during eval by default.
- Autobackend warmup now also warms up Non-Max Suppression (NMS) by running a small synthetic NMS pass after the first forward.
- Updated docstrings to reflect the new `pin_memory` option.

### 🎯 Purpose & Impact
- Faster, smoother first-batch performance by pre-warming NMS kernels on supported devices. 🚀
- More robust memory behavior:
  - Keeps pin_memory on by default where beneficial (GPU training).
  - Disables it during validation to reduce host memory pressure and potential stalls. 🧩
- Backward compatible API: existing code continues to work; advanced users can fine-tune memory behavior.

Example: manually disable pin memory if needed
```python
from ultralytics.data.build import build_dataloader

loader = build_dataloader(dataset, batch=16, workers=8, shuffle=True, pin_memory=False)
```

### CODE REVIEW

**Comment by Y-T-G** [comment]
@glenn-jocher This helps with the PyTorch model validation postptocessing latency.

**main**
```
yolo val model=weights/yolo11n.pt rect=False data=coco8.yaml exist_ok batch=1
Ultralytics 8.3.215 🚀 Python-3.11.13 torch-2.8.0+cu128 CUDA:0 (NVIDIA A100-SXM4-80GB, 81154MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1767.1±603.7 MB/s, size: 54.0 KB)
val: Scanning /datasets/coco8/labels/val.cache... 4 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 4/4 97.0Kit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 13.0it/s 0.3s
                   all          4         17      0.846      0.637       0.85      0.608
                person          3         10          1      0.323      0.603      0.293
                   dog          1          1      0.724          1      0.995      0.697
                 horse          1          2      0.966          1      0.995      0.462
              elephant          1          2      0.667        0.5      0.518      0.307
              umbrella          1          1      0.719          1      0.995      0.995
          potted plant          1          1          1          0      0.995      0.895
Speed: 5.2ms preprocess, 27.9ms inference, 0.0ms loss, 20.5ms postprocess per image
Results saved to /ultralytics/runs/detect/val
💡 Learn more at https://docs.ultralytics.com/modes/val
```

**PR**
```
yolo val model=weights/yolo11n.pt rect=False data=coco8.yaml exist_ok batch=1
Ultralytics 8.3.215 🚀 Python-3.11.13 torch-2.8.0+cu128 CUDA:0 (NVIDIA A100-SXM4-80GB, 81154MiB)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1652.9±689.7 MB/s, size: 54.0 KB)
val: Scanning /datasets/coco8/labels/val.cache... 4 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 4/4 112.6Kit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 18.3it/s 0.2s
                   all          4         17      0.846      0.637       0.85      0.608
                person          3         10          1      0.323      0.603      0.293
                   dog          1          1      0.724          1      0.995      0.697
                 horse          1          2      0.966          1      0.995      0.462
              elephant          1          2      0.667        0.5      0.518      0.307
              umbrella          1          1      0.719          1      0.995      0.995
          potted plant          1          1          1          0      0.995      0.895
Speed: 5.8ms preprocess, 25.7ms inference, 0.0ms loss, 1.6ms postprocess per image
Results saved to /ultralytics/runs/detect/val
```


But TensorRT postprocessing is still slow even with the warmup. It is faster than main, but I don't get why it is slower than PyTorch model when everything seems to be exactly the same in the postprocessing. Both are processing PyTorch tensors of same shape on the same device with the same datatype, but TensorRT is slower. It's very weird.

**main**
```
yolo val model=weights/yolo11n.engine rect=False data=coco8.yaml exist_ok batch=1
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Ultralytics 8.3.215 🚀 Python-3.11.13 torch-2.8.0+cu128 CUDA:0 (NVIDIA A100-SXM4-80GB, 81154MiB)
Loading weights/yolo11n.engine for TensorRT inference...
[10/16/2025-16:56:14] [TRT] [I] Loaded engine size: 13 MiB
[10/16/2025-16:56:14] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +18, now: CPU 0, GPU 28 (MiB)
Setting batch=1 input of shape (1, 3, 640, 640)
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1708.7±558.7 MB/s, size: 54.0 KB)
val: Scanning /datasets/coco8/labels/val.cache... 4 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 4/4 124.3Kit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 7.6it/s 0.5s
                   all          4         17      0.849      0.637       0.85       0.61
                person          3         10          1      0.321      0.603      0.293
                   dog          1          1      0.726          1      0.995      0.697
                 horse          1          2      0.969          1      0.995      0.473
              elephant          1          2      0.676        0.5      0.518      0.307
              umbrella          1          1      0.721          1      0.995      0.995
          potted plant          1          1          1          0      0.995      0.895
Speed: 22.4ms preprocess, 5.1ms inference, 0.8ms loss, 45.0ms postprocess per image
Results saved to /ultralytics/runs/detect/val
💡 Learn more at https://docs.ultralytics.com/modes/val
```

**PR**
```
yolo val model=weights/yolo11n.engine rect=False data=coco8.yaml exist_ok batch=1
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Ultralytics 8.3.215 🚀 Python-3.11.13 torch-2.8.0+cu128 CUDA:0 (NVIDIA A100-SXM4-80GB, 81154MiB)
Loading weights/yolo11n.engine for TensorRT inference...
[10/16/2025-16:57:02] [TRT] [I] Loaded engine size: 13 MiB
[10/16/2025-16:57:02] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +18, now: CPU 0, GPU 28 (MiB)
Setting batch=1 input of shape (1, 3, 640, 640)
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1734.6±411.9 MB/s, size: 54.0 KB)
val: Scanning /datasets/coco8/labels/val.cache... 4 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 4/4 98.1Kit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 11.4it/s 0.3s
                   all          4         17      0.849      0.637       0.85       0.61
                person          3         10          1      0.321      0.603      0.293
                   dog          1          1      0.726          1      0.995      0.697
                 horse          1          2      0.969          1      0.995      0.473
              elephant          1          2      0.676        0.5      0.518      0.307
              umbrella          1          1      0.721          1      0.995      0.995
          potted plant          1          1          1          0      0.995      0.895
Speed: 5.8ms preprocess, 4.6ms inference, 0.0ms loss, 12.4ms postprocess per image
Results saved to /ultralytics/runs/detect/val
```

**Comment by Y-T-G** [comment]
The TensorRT slowdown in my tests was due to `plots=True`. Setting it `False`, reduced the postprocessing latency to about 3ms. And then setting `pin_memory` to `False` reduced it to `1.2ms`:

```
yolo val model=weights/yolo11n.engine rect=False data=coco8.yaml exist_ok batch=1 plots=False
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Ultralytics 8.3.216 🚀 Python-3.11.13 torch-2.8.0+cu128 CUDA:0 (NVIDIA A100-SXM4-80GB, 81154MiB)
Loading weights/yolo11n.engine for TensorRT inference...
[10/17/2025-07:18:45] [TRT] [I] Loaded engine size: 13 MiB
[10/17/2025-07:18:45] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +18, now: CPU 0, GPU 28 (MiB)
Setting batch=1 input of shape (1, 3, 640, 640)
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1713.0±653.2 MB/s, size: 54.0 KB)
val: Scanning /datasets/coco8/labels/val.cache... 4 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 4/4 92.7Kit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 4/4 34.6it/s 0.1s
                   all          4         17      0.849      0.637       0.85       0.61
                person          3         10          1      0.321      0.603      0.293
                   dog          1          1      0.726          1      0.995      0.697
                 horse          1          2      0.969          1      0.995      0.473
              elephant          1          2      0.676        0.5      0.518      0.307
              umbrella          1          1      0.721          1      0.995      0.995
          potted plant          1          1          1          0      0.995      0.895
Speed: 1.3ms preprocess, 1.8ms inference, 0.0ms loss, 1.2ms postprocess per image
```

**Comment by Y-T-G** [comment]
**main**
```
yolo benchmark model=yolo11n.pt data=coco128.yaml format="-"
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1820.8±505.7 MB/s, size: 74.0 KB)
val: Scanning /datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 3.7Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 128/128 49.0it/s 2.6s
                   all        128        929      0.667      0.599      0.675       0.51
Speed: 0.3ms preprocess, 16.2ms inference, 0.0ms loss, 1.6ms postprocess per image
Setup complete ✅ (128 CPUs, 1007.7 GB RAM, 171.2/999.5 GB disk)

Benchmarks complete for yolo11n.pt on coco128.yaml at imgsz=640 (5.81s)
Benchmarks legend:  - ✅ Success  - ❎ Export passed but validation failed  - ❌️ Export failed
+-------------------------------------------------------------------------------------------+
|     Format    Status❔   Size (MB)   metrics/mAP50-95(B)   Inference time (ms/im)   FPS   |
+===========================================================================================+
| 1   PyTorch   ✅         5.4         0.5103                16.24                    61.58 |
+-------------------------------------------------------------------------------------------+

💡 Learn more at https://docs.ultralytics.com/modes/benchmark
```

**PR**
```
yolo benchmark model=yolo11n.pt data=coco128.yaml format="-"
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1398.3±438.2 MB/s, size: 45.1 KB)
val: Scanning /datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 128/128 3.8Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 128/128 646.5it/s 0.2s
                   all        128        929      0.667      0.599      0.675       0.51
Speed: 0.8ms preprocess, 14.2ms inference, 0.0ms loss, 0.9ms postprocess per image
Setup complete ✅ (128 CPUs, 1007.7 GB RAM, 171.2/999.5 GB disk)

Benchmarks complete for yolo11n.pt on coco128.yaml at imgsz=640 (5.86s)
Benchmarks legend:  - ✅ Success  - ❎ Export passed but validation failed  - ❌️ Export failed
+-------------------------------------------------------------------------------------------+
|     Format    Status❔   Size (MB)   metrics/mAP50-95(B)   Inference time (ms/im)   FPS   |
+===========================================================================================+
| 1   PyTorch   ✅         5.4         0.5103                14.22                    70.34 |
+-------------------------------------------------------------------------------------------+
```

**Comment by glenn-jocher** [comment]
@Y-T-G nice work! You think `pin_memory` is affecting postprocessing time?

**Comment by glenn-jocher** [comment]
@Y-T-G merged!

**Comment by UltralyticsAssistant** [comment]
Huge thanks for the merge, @Y-T-G — and shout-out to @glenn-jocher for the assist! This [PR #22425](https://github.com/ultralytics/ultralytics/pull/22425) is a fantastic quality-of-life upgrade: configurable `pin_memory` with smart defaults and an NMS warmup that smooths first-iteration latency. “Details make perfection, and perfection is not a detail.” — Leonardo da Vinci. Your attention to these details delivers faster starts, steadier throughput, and better memory control, all while keeping the API backward compatible. This will elevate training and eval workflows across Ultralytics and Ultralytics HUB. 🚀💡

**Comment by Y-T-G** [comment]
@glenn-jocher I think it affects it during the first few iterations. So it's noticeable with `coco8.yaml`, but less noticeable with larger validation set. Maybe it does some work in the background which consumes GPU resource.
